<a href="https://colab.research.google.com/github/Carmen10-171/01MIAR_ABR26/blob/main/NumPy_estudio_guiado_compartible_(2).ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Práctica guiada — NumPy para el MIAR

**Material de estudio compartido entre compañeros del MIAR.**

Este notebook es tu **gimnasio** de NumPy. Reconstruye los conceptos desde abajo, en bloques pequeños, para que dejes de mirar a NumPy como magia y empieces a usarlo con seguridad. Cuando termines, vas a entender por qué NumPy es la base de casi todo lo que vendrá después en el máster (Pandas, scikit-learn, redes neuronales).

## Cómo se usa

Cada bloque tiene **siempre** esta estructura:

1. **Concepto** — qué es y para qué sirve.
2. **Ejemplo** — código corto comentado. Antes de correrlo, **lee y predice** qué imprime.
3. **Tu ejercicio espejo** — mismo patrón, datos distintos. Lo escribes tú.
4. **Validamos** — corres la celda y comparas con la salida esperada.
5. **Tu reflexión** — completas la frase con tus palabras. Eso fija el patrón a largo plazo.

## Reglas del juego

- **No copies el ejemplo.** El espejo es del mismo *patrón* pero con datos distintos. Si copias, no aprendes.
- **Si te trabas más de 5 min**, busca una pista. Si te trabas 10 min, mira la lógica con calma.
- **Cuando algo falla**, primero piensa qué crees que pasó. Diagnosticar es la mitad del aprendizaje.
- **NO te saltes la reflexión.** Escribir el concepto con tus propias palabras es lo que lo fija (efecto de generación).
- **Importa siempre `numpy as np` al inicio de cada bloque de código** — está hecho así a propósito, para que cada celda corra de forma independiente.

## Por qué NumPy

Una lista de Python es flexible pero lenta: cada elemento es un objeto independiente con su propia cabecera. Un `ndarray` de NumPy es un bloque contiguo de memoria con elementos del mismo tipo, y las operaciones se ejecutan en C. Resultado: 10x a 100x más rápido, sintaxis más corta, y la base sobre la que se construyen Pandas, scikit-learn, PyTorch, TensorFlow...

Vamos. 🚀

---

## Nivel 1 — Crear arrays y entender su forma

Lo primero: cómo se construye un `ndarray` y cómo describir su estructura (forma, dimensiones, tipo de dato). Sin esto, todo lo demás se vuelve confuso porque no sabes *qué* tienes entre manos.

### Bloque 1 — Crear arrays desde listas y constructores básicos

**Concepto:** un `ndarray` es la estructura central de NumPy. Hay varias formas de crear uno:

| Constructor             | Qué hace                                              |
| ----------------------- | ----------------------------------------------------- |
| `np.array([...])`       | Convierte una lista (o lista de listas) en `ndarray`. |
| `np.zeros(shape)`       | Array lleno de ceros con la forma indicada.           |
| `np.ones(shape)`        | Array lleno de unos.                                  |
| `np.full(shape, valor)` | Array lleno del valor que tú elijas.                  |
| `np.eye(n)`             | Matriz identidad de tamaño `n` x `n`.                 |

Para forma 1D pasas un entero (`np.zeros(5)`). Para 2D o más, pasas una **tupla** (`np.zeros((3, 4))`).

#### Mi ejemplo

Predice qué imprime antes de correr.

In [ ]:
import numpy as np

# Desde una lista de Python (1D)
a = np.array([10, 20, 30, 40])
print(a)

# Desde una lista de listas (2D)
b = np.array([[1, 2, 3], [4, 5, 6]])
print(b)

# Arrays prefabricados
print(np.zeros(3))            # 1D de 3 ceros
print(np.ones((2, 3)))        # 2x3 lleno de unos
print(np.full((2, 2), 7))     # 2x2 lleno de 7
print(np.eye(3))              # identidad 3x3

[10 20 30 40]
[[1 2 3]
 [4 5 6]]
[0. 0. 0.]
[[1. 1. 1.]
 [1. 1. 1.]]
[[7 7]
 [7 7]]
[[1. 0. 0.]
 [0. 1. 0.]
 [0. 0. 1.]]


#### Tu ejercicio espejo

En la celda de abajo:

1. Crea un array 1D `temperaturas` a partir de la lista `[18.5, 21.0, 19.7, 22.3]` e imprímelo.
2. Crea un array 2D `matriz` de 3 filas y 2 columnas a partir de `[[1, 2], [3, 4], [5, 6]]`. Imprímelo.
3. Crea un array `ceros` de forma `(4, 4)` lleno de ceros e imprímelo.
4. Crea un array `treses` de forma `(2, 5)` lleno del valor `3` (usa `np.full`). Imprímelo.
5. Crea la matriz identidad `id5` de tamaño 5x5 (con `np.eye`). Imprímela.

**Pista:** para 2D, la forma se pasa como tupla: `np.zeros((4, 4))`, no `np.zeros(4, 4)`.

In [ ]:
import numpy as np

# Tu solución aquí.


#### Tu reflexión

Completa la frase con tus palabras (doble click para editar):

> **"`np.array(...)` lo uso cuando ___. La ventaja sobre una lista de Python es ___."**

### Bloque 2 — `arange` y `linspace` (secuencias numéricas)

**Concepto:** los dos constructores más usados para generar **secuencias** de números:

| Constructor                       | Qué hace                                                                                                          |
| --------------------------------- | ----------------------------------------------------------------------------------------------------------------- |
| `np.arange(inicio, fin, paso)`    | Como `range`, pero acepta `float`. Genera desde `inicio` hasta menos de `fin` con el paso indicado.               |
| `np.linspace(inicio, fin, n)`     | Genera `n` valores **equiespaciados** entre `inicio` y `fin`, **incluyendo ambos extremos**.                      |

La diferencia clave: en `arange` tú controlas el **paso** (y el número de elementos sale solo). En `linspace` tú controlas el **número de elementos** (y el paso sale solo). `linspace` es muy útil cuando dibujas funciones matemáticas (necesitas, por ejemplo, 100 puntos entre 0 y 2π).

#### Mi ejemplo

Predice qué imprime antes de correr.

In [ ]:
import numpy as np

# arange con un solo argumento: como range(8)
print(np.arange(8))

# arange con inicio, fin y paso (¡acepta decimales!)
print(np.arange(0, 2, 0.5))

# linspace: 5 valores equiespaciados entre 0 y 1, incluidos
print(np.linspace(0, 1, 5))

# linspace típico de gráficas: 11 puntos entre -1 y 1
print(np.linspace(-1, 1, 11))

[0 1 2 3 4 5 6 7]
[0.  0.5 1.  1.5]
[0.   0.25 0.5  0.75 1.  ]
[-1.  -0.8 -0.6 -0.4 -0.2  0.   0.2  0.4  0.6  0.8  1. ]


#### Tu ejercicio espejo

En la celda de abajo:

1. Usa `np.arange` para crear un array con los enteros del **0 al 9** (no incluyas el 10). Imprímelo.
2. Usa `np.arange` para crear un array con los **múltiplos de 5 entre 0 y 50** (sin incluir el 50). Esperado: `[ 0  5 10 15 20 25 30 35 40 45]`.
3. Usa `np.arange` con paso decimal para generar valores desde `1.0` hasta menos de `3.0` con paso `0.25`. Imprímelo.
4. Usa `np.linspace` para generar **7 valores equiespaciados entre 0 y 12** (ambos incluidos). Imprímelo.
5. Usa `np.linspace` para generar **100 valores entre 0 y 2π** (usa `np.pi`). No imprimas los 100, pero sí imprime su **longitud** con `len(...)` para confirmar que tiene 100 elementos.

**Pista:** la constante π está en `np.pi`.

In [ ]:
import numpy as np

# Tu solución aquí.


#### Tu reflexión

Completa la frase con tus palabras (doble click para editar):

> **"`np.arange` lo uso cuando quiero controlar ___, y `np.linspace` cuando quiero controlar ___. A diferencia de `range`, `np.arange` sí acepta ___."**

### Bloque 3 — Atributos: `shape`, `ndim`, `size`, `dtype`

**Concepto:** todo `ndarray` lleva consigo cuatro atributos que lo describen:

| Atributo  | Qué te dice                                                          |
| --------- | -------------------------------------------------------------------- |
| `shape`   | La forma como tupla. Ej: `(3, 4)` significa 3 filas y 4 columnas.    |
| `ndim`    | Número de dimensiones (1 para 1D, 2 para matriz, 3 para tensor 3D…). |
| `size`    | Cantidad **total** de elementos (producto de `shape`).               |
| `dtype`   | Tipo de cada elemento (`int64`, `float64`, `bool`, etc.).            |

**Detalle importante**: a diferencia de las listas de Python, todos los elementos de un `ndarray` son del mismo `dtype`. Si mezclas `int` y `float`, NumPy promociona todo a `float`.

Estos atributos no son funciones — no llevan paréntesis (`a.shape`, no `a.shape()`).

#### Mi ejemplo

Predice qué imprime antes de correr.

In [ ]:
import numpy as np

# Array 1D con 4 enteros
a = np.array([10, 20, 30, 40])
print(a.shape, a.ndim, a.size, a.dtype)

# Array 2D de 2 filas x 3 columnas
b = np.array([[1, 2, 3], [4, 5, 6]])
print(b.shape, b.ndim, b.size, b.dtype)

# Promoción de tipo: si mezclo int y float, todo se vuelve float
c = np.array([1, 2, 3.5])
print(c.dtype)

# Tensor 3D: 2 'capas' de 3 filas x 4 columnas
d = np.zeros((2, 3, 4))
print(d.shape, d.ndim, d.size)

(4,) 1 4 int64
(2, 3) 2 6 int64
float64
(2, 3, 4) 3 24


#### Tu ejercicio espejo

Datos de partida (no los modifiques, solo úsalos):

```python
x = np.arange(12)
y = np.ones((4, 5))
z = np.full((2, 3, 4), 7)
```

En la celda de abajo:

1. Crea los tres arrays `x`, `y`, `z` como se indica arriba.
2. Para cada uno, imprime `shape`, `ndim`, `size` y `dtype` en una línea con f-string. Ejemplo de salida para `x`: `x: shape=(12,), ndim=1, size=12, dtype=int64`.
3. Crea un array `mezcla = np.array([1, 2, 3.0, 4])` (un solo float entre enteros) e imprime su `dtype`. ¿Promocionó o no?
4. Calcula tú mismo, **antes de imprimir `z.size`**, cuántos elementos debe tener `z` (basta multiplicar los números de la forma). Luego confirma con el `print`.

**Pista:** no pongas paréntesis en los atributos. `x.shape`, no `x.shape()`.

In [ ]:
import numpy as np

# Tu solución aquí.


#### Tu reflexión

Completa la frase con tus palabras (doble click para editar):

> **"`shape` me dice ___, `ndim` me dice ___, `size` me dice ___, y `dtype` me dice ___. Todos los elementos de un `ndarray` deben tener ___."**

---

## Nivel 2 — Indexar y trocear (slicing)

Ya sabes crear arrays. Ahora vas a aprender a **acceder** a partes concretas: un elemento, una fila, una columna, una submatriz, o un conjunto de elementos que cumplen una condición. Esta es la operación más usada en el día a día con NumPy.

### Bloque 4 — Indexar y trocear arrays 1D

**Concepto:** indexar un array 1D funciona igual que indexar una lista de Python.

- `a[0]` — primer elemento.
- `a[-1]` — último elemento.
- `a[2:5]` — elementos en posiciones 2, 3, 4 (sin incluir el 5).
- `a[:3]` — los tres primeros.
- `a[::2]` — todos con paso 2.
- `a[::-1]` — al revés.

**Diferencia clave con las listas:** el slice en una lista te da una **copia** nueva, pero el slice en un `ndarray` te da una **vista** (referencia al mismo bloque de memoria). Si modificas el slice, modificas el original. Lo veremos a fondo en el Bloque 6.

#### Mi ejemplo

Predice qué imprime antes de correr.

In [ ]:
import numpy as np

a = np.arange(10, 20)   # [10 11 12 13 14 15 16 17 18 19]
print(a)

# Acceso por índice
print(a[0])             # primer elemento
print(a[-1])            # último

# Slices
print(a[2:5])           # del 2 al 4 (sin el 5)
print(a[:4])            # los 4 primeros
print(a[6:])            # desde el 6 hasta el final
print(a[::2])           # paso 2
print(a[::-1])          # invertido

[10 11 12 13 14 15 16 17 18 19]
10
19
[12 13 14]
[10 11 12 13]
[16 17 18 19]
[10 12 14 16 18]
[19 18 17 16 15 14 13 12 11 10]


#### Tu ejercicio espejo

Datos:

```python
v = np.arange(100, 110)   # [100 101 102 ... 109]
```

En la celda de abajo:

1. Crea `v` y imprímelo.
2. Imprime el **tercer** elemento y el **penúltimo** (usando índice negativo).
3. Imprime los **4 últimos** elementos con un slice.
4. Imprime los elementos en posiciones **pares** (0, 2, 4, 6, 8) usando un slice con paso.
5. Imprime el array **al revés**.

**Pista:** penúltimo es `v[-2]`. Los 4 últimos son `v[-4:]`.

In [ ]:
import numpy as np

# Tu solución aquí.


#### Tu reflexión

Completa la frase con tus palabras (doble click para editar):

> **"En un array 1D, `a[i]` me da ___, `a[-1]` me da ___, y `a[i:j]` me da los elementos desde ___ hasta ___ (sin incluir j). El paso se pone en ___."**

### Bloque 5 — Indexar y trocear arrays 2D

**Concepto:** en 2D usamos **dos índices separados por coma**: `a[fila, columna]`. Cada índice puede ser un número o un slice. Esta sintaxis es la diferencia más grande con las listas anidadas de Python.

| Expresión       | Qué obtienes                                                |
| --------------- | ----------------------------------------------------------- |
| `a[1, 2]`       | El elemento en fila 1, columna 2 (un escalar).              |
| `a[1, :]`       | Toda la fila 1 (vector 1D).                                 |
| `a[:, 0]`       | Toda la columna 0 (vector 1D).                              |
| `a[1:3, :]`     | Las filas 1 y 2 enteras.                                    |
| `a[:, 1:3]`     | Las columnas 1 y 2 enteras.                                 |
| `a[1:3, 0:2]`   | Subbloque: filas 1-2, columnas 0-1.                         |

El `:` solo significa "todos". Tener `a[1, :]` y `a[1]` da el mismo resultado, pero el `:` explícito hace el código más legible.

#### Mi ejemplo

Predice qué imprime antes de correr.

In [ ]:
import numpy as np

# Matriz 4x4 con valores del 0 al 15
m = np.arange(16).reshape(4, 4)
print(m)

print("---")

# Un elemento (fila 1, columna 2)
print(m[1, 2])

# Una fila entera
print(m[2, :])

# Una columna entera
print(m[:, 0])

# Submatriz: filas 1 y 2, columnas 1 y 2
print(m[1:3, 1:3])

[[ 0  1  2  3]
 [ 4  5  6  7]
 [ 8  9 10 11]
 [12 13 14 15]]
---
6
[ 8  9 10 11]
[ 0  4  8 12]
[[ 5  6]
 [ 9 10]]


#### Tu ejercicio espejo

Datos:

```python
M = np.arange(1, 26).reshape(5, 5)   # 5x5 con valores del 1 al 25
```

En la celda de abajo:

1. Crea `M` e imprímelo (echa un vistazo a la disposición para hacerte una imagen mental).
2. Imprime el elemento en la **fila 2, columna 3** (recuerda que las posiciones empiezan en 0).
3. Imprime la **primera columna** entera.
4. Imprime la **última fila** entera (usa índice negativo).
5. Imprime la **submatriz central 3x3** (filas 1-3, columnas 1-3).

**Pista:** para la submatriz central, recuerda que el slice no incluye el extremo derecho: `1:4` te da las filas 1, 2 y 3.

In [ ]:
import numpy as np

# Tu solución aquí.


#### Tu reflexión

Completa la frase con tus palabras (doble click para editar):

> **"En un array 2D, el primer índice es ___ y el segundo es ___. Para sacar una fila entera escribo ___ y para sacar una columna entera escribo ___."**

### Bloque 6 — Vista vs copia (la trampa silenciosa)

**Concepto:** en Python puro, `lista[2:5]` devuelve una **copia** nueva. En NumPy, `array[2:5]` devuelve una **vista** — una referencia al mismo bloque de memoria. Si modificas la vista, modificas el original. Si no eres consciente, esto produce bugs muy difíciles de detectar.

Resumen:

| Operación                          | ¿Vista o copia?                 |
| ---------------------------------- | ------------------------------- |
| Slice básico (`a[2:5]`)            | **Vista**                       |
| Indexar con un escalar (`a[3]`)    | Copia (devuelve un escalar)     |
| Boolean mask (`a[a > 0]`)          | **Copia** (lo verás en Bloque 8)|
| Fancy indexing (`a[[0, 2, 4]]`)    | **Copia** (lo verás en Bloque 9)|
| `a.copy()`                         | Copia explícita garantizada     |

Regla práctica: **si quieres asegurarte de no tocar el original, llama a `.copy()` explícitamente.**

#### Mi ejemplo

Predice qué imprime antes de correr. Especialmente: ¿qué le pasa a `original` cuando modifico `vista`?

In [ ]:
import numpy as np

# Caso 1: con lista de Python — el slice devuelve copia
lista = [0, 1, 2, 3, 4]
trozo = lista[1:4]
trozo[0] = 999
print(lista)        # ¿se modificó la lista original?
print(trozo)

print("---")

# Caso 2: con ndarray — el slice devuelve vista
original = np.arange(5)
vista = original[1:4]
vista[0] = 999
print(original)     # ¿qué pasó aquí?
print(vista)

print("---")

# Caso 3: forzar copia con .copy()
original2 = np.arange(5)
copia = original2[1:4].copy()
copia[0] = 999
print(original2)    # ¿ahora?
print(copia)

[0, 1, 2, 3, 4]
[999, 2, 3]
---
[  0 999   2   3   4]
[999   2   3]
---
[0 1 2 3 4]
[999   2   3]


#### Tu ejercicio espejo

En la celda de abajo:

1. Crea `datos = np.arange(20, 30)` e imprímelo.
2. Crea una vista `mitad = datos[3:7]` y modifica su primer elemento: `mitad[0] = -1`.
3. Imprime `datos` después de la modificación. **Anota mentalmente** qué pasó.
4. Crea ahora `datos2 = np.arange(20, 30)` y haz `copia = datos2[3:7].copy()`. Modifica `copia[0] = -1`.
5. Imprime `datos2` y `copia`. Compara con el resultado anterior.

**Reto extra:** ¿qué crees que pasa si el slice abarca toda la matriz (`datos[:]`)? ¿Es vista o copia? Pruébalo y compruébalo.

**Pista:** este es exactamente el tipo de error que aparece cuando crees que estás modificando datos "aislados" y de pronto se rompe otra cosa en tu pipeline.

In [ ]:
import numpy as np

# Tu solución aquí.


#### Tu reflexión

Completa la frase con tus palabras (doble click para editar):

> **"Un slice de NumPy devuelve una ___, no una copia. Si modifico el resultado, también se modifica ___. Para evitarlo uso ___."**

### Bloque 7 — Boolean masking (filtrado vectorizado)

**Concepto:** este es uno de los superpoderes de NumPy. Una **máscara booleana** es un array de `True`/`False` del mismo tamaño que tu array de datos. Cuando indexas un array con una máscara, te quedas con los elementos donde la máscara es `True`.

Lo brutal: **las comparaciones vectorizadas (`a > 0`, `a == 5`, etc.) ya devuelven una máscara**. Así que filtrar elementos pasa de un bucle de varias líneas a una sola línea.

Comparación con Python puro:

```python
# Python puro: filtrar positivos
positivos = []
for x in lista:
    if x > 0:
        positivos.append(x)

# NumPy: lo mismo en una línea
positivos = arr[arr > 0]
```

Y además puedes **asignar valores** a través de la máscara (`arr[arr < 0] = 0` cambia todos los negativos a 0 de golpe).

#### Mi ejemplo

Predice qué imprime antes de correr.

In [ ]:
import numpy as np

a = np.array([-3, -1, 0, 2, 5, -7, 4, 1])
print(a)

# Una comparación vectorizada genera la máscara
mascara = a > 0
print(mascara)

# Aplicar la máscara: quedarme solo con los positivos
print(a[mascara])

# Lo mismo en una sola línea
print(a[a > 0])

print("---")

# Asignar a través de la máscara: cambiar los negativos a 0
b = a.copy()
b[b < 0] = 0
print(b)

# Combinar condiciones: & en lugar de 'and', | en lugar de 'or'
# (¡y poner cada comparación entre paréntesis!)
print(a[(a > 0) & (a < 5)])

[-3 -1  0  2  5 -7  4  1]
[False False False  True  True False  True  True]
[2 5 4 1]
[2 5 4 1]
---
[0 0 0 2 5 0 4 1]
[2 4 1]


#### Tu ejercicio espejo

Datos:

```python
notas = np.array([4, 7, 5, 9, 3, 8, 6, 10, 2, 5])
```

En la celda de abajo:

1. Crea `notas` e imprímelo.
2. Imprime la **máscara booleana** que marca con `True` las notas **aprobadas** (>= 5).
3. Imprime solo las **notas aprobadas** usando esa máscara.
4. Imprime las notas que están entre 5 y 8 **ambos inclusive**. Usa `&` y paréntesis.
5. Crea una copia `notas2 = notas.copy()` y pon a 5 todas las que sean menores que 5 (los suspensos se aprueban por la cara). Imprime `notas2`.

**Pista para el punto 4:** `(notas >= 5) & (notas <= 8)`. Los paréntesis son **obligatorios**, sin ellos NumPy se confunde con la precedencia.

In [ ]:
import numpy as np

# Tu solución aquí.


#### Tu reflexión

Completa la frase con tus palabras (doble click para editar):

> **"Una máscara booleana es ___. Si la uso con `a[mascara]` me quedo con ___. Para combinar condiciones uso ___ (and) y ___ (or), siempre entre paréntesis."**

### Bloque 8 — Fancy indexing (elegir por lista de posiciones)

**Concepto:** además de slices y máscaras, NumPy te deja indexar pasándole **una lista de índices**. El resultado es un array nuevo con esos elementos en ese orden.

Esto es brutalmente útil para:

- Reordenar elementos.
- Seleccionar filas o columnas concretas (no necesariamente contiguas).
- Tomar muestras aleatorias.

**Importante:** el fancy indexing devuelve **una copia**, no una vista (al contrario que el slice básico del Bloque 6).

#### Mi ejemplo

Predice qué imprime antes de correr.

In [ ]:
import numpy as np

a = np.array([100, 200, 300, 400, 500])

# Elegir varios elementos a la vez con una lista de índices
print(a[[0, 2, 4]])

# Puedo repetir índices y reordenar
print(a[[4, 4, 0, 1]])

print("---")

# En 2D: elegir filas concretas
m = np.arange(20).reshape(5, 4)
print(m)
print(m[[0, 2, 4]])           # filas 0, 2 y 4
print(m[:, [0, 3]])           # columnas 0 y 3 de todas las filas

[100 300 500]
[500 500 100 200]
---
[[ 0  1  2  3]
 [ 4  5  6  7]
 [ 8  9 10 11]
 [12 13 14 15]
 [16 17 18 19]]
[[ 0  1  2  3]
 [ 8  9 10 11]
 [16 17 18 19]]
[[ 0  3]
 [ 4  7]
 [ 8 11]
 [12 15]
 [16 19]]


#### Tu ejercicio espejo

Datos:

```python
ciudades = np.array(["Madrid", "Bilbao", "Sevilla", "Valencia", "Málaga", "Granada"])
matriz = np.arange(30).reshape(6, 5)
```

En la celda de abajo:

1. Crea `ciudades` y `matriz`, e imprímelos.
2. Usa fancy indexing para imprimir las ciudades en las **posiciones 0, 3 y 5**.
3. Imprime las ciudades **en orden inverso al original** (de la 5 a la 0) usando una lista de índices `[5, 4, 3, 2, 1, 0]`.
4. De `matriz`, imprime las **filas 1, 3 y 5** (las impares).
5. De `matriz`, imprime las **columnas 0, 2 y 4** (las pares) — **de todas las filas**.

**Pista para el punto 5:** `matriz[:, [0, 2, 4]]`.

In [ ]:
import numpy as np

# Tu solución aquí.


#### Tu reflexión

Completa la frase con tus palabras (doble click para editar):

> **"Fancy indexing lo uso cuando quiero seleccionar ___ elementos por sus posiciones. A diferencia del slice básico, devuelve una ___ (no una vista)."**

---

## Nivel 3 — Operaciones vectorizadas

Aquí ves la gran ventaja de NumPy: **operar sobre todo el array a la vez**, sin escribir bucles. La sintaxis se parece a la del álgebra de toda la vida, y por debajo se ejecuta en C, así que es 10x-100x más rápido que un `for` de Python.

### Bloque 9 — Aritmética elemento a elemento

**Concepto:** al sumar/restar/multiplicar/dividir dos arrays del mismo tamaño, NumPy aplica la operación **elemento por elemento**. Y si uno de los operandos es un escalar, lo aplica a cada elemento del array.

Comparación clave con listas de Python:

```python
[1, 2, 3] * 3        # → [1, 2, 3, 1, 2, 3, 1, 2, 3]   (repite la lista)
np.array([1, 2, 3]) * 3   # → array([3, 6, 9])         (multiplica cada elemento)
```

**Esta diferencia es la primera fuente de confusión** cuando vienes de Python puro. Para listas, `*` repite. Para `ndarray`, `*` multiplica elemento a elemento.

Y lo mejor: las operaciones vectorizadas son **mucho más rápidas** que un bucle equivalente en Python. Vamos a medirlo.

#### Mi ejemplo

Predice qué imprime antes de correr.

In [ ]:
import numpy as np
import time

a = np.array([1, 2, 3, 4])
b = np.array([10, 20, 30, 40])

# Aritmética entre dos arrays del mismo tamaño
print(a + b)
print(b - a)
print(a * b)
print(b / a)

print("---")

# Aritmética con escalar
print(a * 10)
print(a ** 2)        # cuadrado de cada elemento

print("---")

# Comparación de velocidad con listas Python
n = 1_000_000
lst = list(range(n))
arr = np.arange(n)

t0 = time.time()
lst2 = [x * 2 for x in lst]
print(f"Lista Python: {time.time() - t0:.4f} s")

t0 = time.time()
arr2 = arr * 2
print(f"NumPy:        {time.time() - t0:.4f} s")

[11 22 33 44]
[ 9 18 27 36]
[ 10  40  90 160]
[10. 10. 10. 10.]
---
[10 20 30 40]
[ 1  4  9 16]
---
Lista Python: 0.3600 s
NumPy:        0.0075 s


#### Tu ejercicio espejo

Datos:

```python
precios = np.array([10.0, 25.0, 40.0, 8.0, 100.0])
cantidades = np.array([2, 1, 3, 5, 1])
```

En la celda de abajo:

1. Crea ambos arrays. Calcula `subtotal = precios * cantidades` (multiplicación elemento a elemento) e imprímelo.
2. Aplica un **IVA del 21%** a todos los precios sin bucles: `precios_con_iva = precios * 1.21`. Imprime.
3. Calcula el **descuento de 2 euros por cada producto**: `precios_dto = precios - 2`. Imprime.
4. Calcula el **cuadrado** de cada cantidad (`cantidades ** 2`). Imprime.
5. **Bonus — medir velocidad.** Genera dos arrays grandes (1 millón de elementos) con `np.arange` y multiplícalos elemento a elemento. Hazlo también con listas y un `for`. Mide el tiempo con `time.time()` y compara.

**Pista:** importa también `import time` para el punto 5.

In [ ]:
import numpy as np
import time

# Tu solución aquí.


#### Tu reflexión

Completa la frase con tus palabras (doble click para editar):

> **"Cuando hago `a + b` con dos `ndarray`s, NumPy suma ___. Si uno es un escalar, lo aplica a ___. Es más rápido que un bucle porque ___."**

### Bloque 10 — Funciones universales (ufuncs): `sqrt`, `exp`, `sin`, `log`...

**Concepto:** las *ufuncs* son funciones de NumPy que se aplican **elemento a elemento** a todo un array. La sintaxis es siempre `np.funcion(array)` y devuelven otro array de la misma forma.

Las más usadas en el día a día:

| Función       | Qué hace                                            |
| ------------- | --------------------------------------------------- |
| `np.sqrt(a)`  | Raíz cuadrada de cada elemento.                     |
| `np.exp(a)`   | e elevado a cada elemento.                          |
| `np.log(a)`   | Logaritmo natural (en base e).                      |
| `np.log2(a)`  | Logaritmo en base 2.                                |
| `np.sin(a)`   | Seno de cada elemento (en radianes).                |
| `np.cos(a)`   | Coseno de cada elemento.                            |
| `np.abs(a)`   | Valor absoluto.                                     |
| `np.round(a)` | Redondear cada elemento.                            |

Si en Python puro escribirías `[math.sqrt(x) for x in lista]`, en NumPy es solo `np.sqrt(arr)`.

#### Mi ejemplo

Predice qué imprime antes de correr.

In [ ]:
import numpy as np

a = np.array([1, 4, 9, 16, 25])
print(np.sqrt(a))                # raíces cuadradas

b = np.array([0, 1, 2, 3])
print(np.exp(b))                 # e^0, e^1, e^2, e^3

c = np.linspace(0, np.pi, 5)     # 5 puntos entre 0 y pi
print(c)
print(np.sin(c))                 # seno de cada uno (al final, ~0)

d = np.array([-3.5, -1.2, 0.4, 2.7])
print(np.abs(d))                 # valor absoluto
print(np.round(d))               # redondeado al entero más cercano

[1. 2. 3. 4. 5.]
[ 1.          2.71828183  7.3890561  20.08553692]
[0.         0.78539816 1.57079633 2.35619449 3.14159265]
[0.00000000e+00 7.07106781e-01 1.00000000e+00 7.07106781e-01
 1.22464680e-16]
[3.5 1.2 0.4 2.7]
[-4. -1.  0.  3.]


#### Tu ejercicio espejo

En la celda de abajo:

1. Crea `n = np.arange(1, 11)` (números del 1 al 10) e imprime su **raíz cuadrada**.
2. Crea `x = np.linspace(0, 2 * np.pi, 9)` (9 puntos entre 0 y 2π) e imprime el **seno** y el **coseno** de cada uno.
3. Crea `p = np.array([1, 10, 100, 1000, 10000])` y calcula su **logaritmo en base 10** (`np.log10`). Esperado: `[0. 1. 2. 3. 4.]`.
4. Crea `mezcla = np.array([-5.7, -2.1, 0.0, 3.4, 8.9])` y calcula `np.abs(mezcla)`. Después aplícale `np.round(mezcla)`.
5. **Bonus:** calcula la fórmula del módulo (norma) de un vector 2D: dados `vx = 3` y `vy = 4`, el módulo es `sqrt(vx^2 + vy^2)`. Pero hazlo para **muchos vectores a la vez**: crea `vx = np.array([3, 6, 5, 8])` y `vy = np.array([4, 8, 12, 15])` y calcula sus módulos en una sola línea. Esperado: `[ 5. 10. 13. 17.]`.

**Pista para el bonus:** `np.sqrt(vx**2 + vy**2)`.

In [ ]:
import numpy as np

# Tu solución aquí.


#### Tu reflexión

Completa la frase con tus palabras (doble click para editar):

> **"Las ufuncs (`np.sqrt`, `np.exp`, `np.sin`...) se aplican ___ a todo el array. Me ahorran tener que escribir ___ para aplicar la función a cada elemento."**

---

## Nivel 4 — Reducciones y agregaciones

Las reducciones colapsan un array en **un número** (o en un array más pequeño): suma, media, mínimo, máximo, desviación típica... Son la base de cualquier análisis estadístico, y dominan los problemas reales de datos.

### Bloque 11 — `sum`, `mean`, `std`, `min`, `max`, `argmin`, `argmax`

**Concepto:** una **reducción** toma un array y devuelve un único valor que lo resume. Las más usadas:

| Función           | Qué devuelve                                                      |
| ----------------- | ----------------------------------------------------------------- |
| `a.sum()`         | Suma de todos los elementos.                                      |
| `a.mean()`        | Media aritmética.                                                 |
| `a.std()`         | Desviación típica.                                                |
| `a.min()`         | Valor mínimo.                                                     |
| `a.max()`         | Valor máximo.                                                     |
| `a.argmin()`      | **Índice** del mínimo (en qué posición está).                     |
| `a.argmax()`      | **Índice** del máximo.                                            |
| `a.cumsum()`      | Suma acumulada (devuelve un array, no un escalar).                |

Todas las puedes invocar como método (`a.mean()`) o como función de NumPy (`np.mean(a)`). Funcionan igual.

La diferencia clave entre `min`/`max` y `argmin`/`argmax`: las primeras te dan el **valor**, las segundas te dan **dónde está** ese valor. Esto es muy útil cuando, por ejemplo, quieres saber **qué día** hizo más calor (no cuántos grados — eso ya lo sabes — sino la posición).

#### Mi ejemplo

Predice qué imprime antes de correr.

In [ ]:
import numpy as np

ventas = np.array([100, 250, 80, 410, 150, 320, 90])

print("Suma:",  ventas.sum())
print("Media:", ventas.mean())
print("Std:",   ventas.std())
print("Min:",   ventas.min())
print("Max:",   ventas.max())

print("---")

# argmin / argmax: índice donde está el valor extremo
print("Índice del mínimo:", ventas.argmin())
print("Índice del máximo:", ventas.argmax())

# Suma acumulada (cada elemento es la suma de todos los anteriores y él mismo)
print("Cumsum:", ventas.cumsum())

Suma: 1400
Media: 200.0
Std: 119.52286093343936
Min: 80
Max: 410
---
Índice del mínimo: 2
Índice del máximo: 3
Cumsum: [ 100  350  430  840  990 1310 1400]


#### Tu ejercicio espejo

Datos (temperaturas máximas de una semana, en °C):

```python
temperaturas = np.array([22.5, 24.0, 19.8, 28.1, 26.4, 21.0, 25.7])
dias = np.array(["Lun", "Mar", "Mié", "Jue", "Vie", "Sáb", "Dom"])
```

En la celda de abajo:

1. Crea ambos arrays. Imprime la **temperatura media de la semana** redondeada a 2 decimales.
2. Imprime la **desviación típica** redondeada a 2 decimales.
3. Imprime el **día más caluroso** (no la temperatura, sino el nombre del día). Pista: usa `argmax` para obtener el índice y luego indexa `dias`.
4. Imprime el **día más frío** de la misma forma.
5. Imprime la **suma acumulada** de temperaturas. Esto representa la "temperatura total acumulada" día a día.

**Pista para 3 y 4:** `dias[temperaturas.argmax()]`.

In [ ]:
import numpy as np

# Tu solución aquí.


#### Tu reflexión

Completa la frase con tus palabras (doble click para editar):

> **"Las reducciones colapsan un array en ___. `min`/`max` me dan ___ y `argmin`/`argmax` me dan ___. Combinándolos puedo saber dónde está el valor extremo."**

### Bloque 12 — Reducciones por eje (`axis=0`, `axis=1`)

**Concepto:** cuando trabajas con una matriz 2D, **no siempre quieres colapsarlo todo a un único número**. Muchas veces lo que quieres es resumir por filas o por columnas. Para eso usas el argumento `axis`.

Regla de oro (la causa #1 de confusión en NumPy):

- `axis=0` significa "colapsa las **filas**, dame un resultado **por columna**".
- `axis=1` significa "colapsa las **columnas**, dame un resultado **por fila**".

Otra forma de verlo: `axis` indica el eje que **desaparece** en el resultado.

Imagina una matriz de ventas con **filas = tiendas** y **columnas = productos**:

- `ventas.sum(axis=0)` → total vendido **por producto** (suma todas las tiendas).
- `ventas.sum(axis=1)` → total vendido **por tienda** (suma todos los productos).

#### Mi ejemplo

Predice qué imprime antes de correr.

In [ ]:
import numpy as np

# Matriz 3 tiendas (filas) x 4 productos (columnas)
ventas = np.array([
    [10, 20, 30, 40],   # tienda 0
    [ 5, 15, 25, 35],   # tienda 1
    [50, 60, 70, 80],   # tienda 2
])
print(ventas)

print("---")

# Total sin axis: todo el array a un único número
print("Total global:", ventas.sum())

# axis=0 → resultado por columna (un número por cada producto)
print("Total por producto:", ventas.sum(axis=0))

# axis=1 → resultado por fila (un número por cada tienda)
print("Total por tienda:", ventas.sum(axis=1))

print("---")

# Media de cada producto a lo largo de las tiendas
print("Media por producto:", ventas.mean(axis=0))

# Producto más vendido EN CADA TIENDA (índice de la columna)
print("Producto top por tienda:", ventas.argmax(axis=1))

[[10 20 30 40]
 [ 5 15 25 35]
 [50 60 70 80]]
---
Total global: 440
Total por producto: [ 65  95 125 155]
Total por tienda: [100  80 260]
---
Media por producto: [21.66666667 31.66666667 41.66666667 51.66666667]
Producto top por tienda: [3 3 3]


#### Tu ejercicio espejo

Datos (notas de 4 alumnos en 3 asignaturas):

```python
notas = np.array([
    [7, 8, 6],   # Alumno 0
    [5, 9, 4],   # Alumno 1
    [10, 7, 8],  # Alumno 2
    [3, 6, 7],   # Alumno 3
])
alumnos = np.array(["Ana", "Beto", "Carla", "Diego"])
asignaturas = np.array(["Mate", "Lengua", "Historia"])
```

En la celda de abajo:

1. Crea los tres arrays. Imprime la **media de cada alumno** (un valor por fila → `axis=1`).
2. Imprime la **media de cada asignatura** (un valor por columna → `axis=0`).
3. Imprime la **nota máxima de cada alumno** y, usando `argmax(axis=1)`, **en qué asignatura sacó esa máxima**. Pista: indexa `asignaturas` con el resultado.
4. Imprime el **alumno con mejor media**. Pasos: calcula medias por alumno → `argmax` para el índice del mejor → indexa `alumnos`.
5. Imprime cuántos **suspensos** (notas < 5) tiene **cada alumno**. Pista: `(notas < 5).sum(axis=1)`. Es un truco brutal: `True` cuenta como 1 al sumar.

**Pista clave del 5:** una máscara booleana (`notas < 5`) sumada con `axis` te cuenta cuántos `True` hay por fila/columna.

In [ ]:
import numpy as np

# Tu solución aquí.


#### Tu reflexión

Completa la frase con tus palabras (doble click para editar):

> **"`axis=0` me da un resultado por ___ (colapsa las filas). `axis=1` me da un resultado por ___ (colapsa las columnas). Sin `axis`, la reducción colapsa ___."**

---

## Nivel 5 — Reshape, transpuesta y broadcasting

El paso de mejor relación esfuerzo/beneficio en NumPy: aprender a cambiar la **forma** de los arrays y entender cómo NumPy combina arrays de formas distintas (broadcasting). Cuando lo asimilas, tu código se vuelve corto, claro y eficiente.

### Bloque 13 — `reshape`, `flatten`, `T` (transpuesta)

**Concepto:** un mismo bloque de datos puede verse con distintas formas. **`reshape` cambia la forma sin tocar los datos.**

| Operación             | Qué hace                                                                                |
| --------------------- | --------------------------------------------------------------------------------------- |
| `a.reshape(filas, cols)` | Devuelve una vista con la forma indicada. El producto de las dimensiones debe coincidir con el total de elementos. |
| `a.reshape(filas, -1)`   | Pones `-1` en una dimensión y NumPy la calcula sola. Muy práctico.                  |
| `a.flatten()`         | Aplana a 1D (siempre copia).                                                            |
| `a.T`                 | Transpuesta — filas y columnas intercambiadas. No lleva paréntesis.                     |

Regla obvia pero crítica: para hacer `reshape(3, 4)` el array tiene que tener exactamente **12 elementos** (3 × 4). Si no, NumPy te tira `ValueError`.

#### Mi ejemplo

Predice qué imprime antes de correr.

In [ ]:
import numpy as np

a = np.arange(12)
print(a)
print(a.shape)

print("---")

# Reshape a 3 filas x 4 columnas
m = a.reshape(3, 4)
print(m)
print(m.shape)

# Reshape con -1: dejo que NumPy calcule
n = a.reshape(2, -1)    # 2 filas, las columnas que toquen → 6
print(n)
print(n.shape)

print("---")

# Aplanar de vuelta a 1D
print(m.flatten())

# Transpuesta (filas <-> columnas)
print(m.T)
print(m.T.shape)        # antes era (3, 4), ahora (4, 3)

[ 0  1  2  3  4  5  6  7  8  9 10 11]
(12,)
---
[[ 0  1  2  3]
 [ 4  5  6  7]
 [ 8  9 10 11]]
(3, 4)
[[ 0  1  2  3  4  5]
 [ 6  7  8  9 10 11]]
(2, 6)
---
[ 0  1  2  3  4  5  6  7  8  9 10 11]
[[ 0  4  8]
 [ 1  5  9]
 [ 2  6 10]
 [ 3  7 11]]
(4, 3)


#### Tu ejercicio espejo

En la celda de abajo:

1. Crea `a = np.arange(1, 25)` (24 elementos, del 1 al 24).
2. Conviértelo en una matriz de **4 filas y 6 columnas** con `reshape`. Imprime forma y matriz.
3. Conviértelo en una matriz de **6 filas y... la que toque**, usando `-1`. Imprime forma y matriz.
4. Toma la matriz 4x6 y calcula su **transpuesta**. Imprime forma y matriz transpuesta (debe ser 6x4).
5. **Aplana** la matriz transpuesta y compara con el array original. ¿Salen iguales o no? ¿Por qué crees que pasa eso?

**Pista para el punto 5:** la transpuesta recorre los datos en un orden distinto (columna a columna en lugar de fila a fila). Si los aplanas, NO obtendrás el orden original.

In [ ]:
import numpy as np

# Tu solución aquí.


#### Tu reflexión

Completa la frase con tus palabras (doble click para editar):

> **"`reshape(filas, cols)` lo uso cuando quiero ___ sin perder datos. El producto filas × cols debe ser igual al ___. `T` me da ___, intercambiando filas y columnas."**

### Bloque 14 — Broadcasting (arrays de distinto tamaño que aun así se combinan)

**Concepto:** broadcasting es el sistema que tiene NumPy para combinar arrays de **formas distintas** sin que tengas que escribir bucles. Ya lo usaste implícitamente cuando hiciste `arr * 2` (un array y un escalar). Pero también funciona entre arrays.

Reglas (mira con calma, son lógicas):

1. NumPy compara las formas **desde la derecha hacia la izquierda**.
2. Dos dimensiones son compatibles si son iguales **o** una de ellas es 1.
3. Si una dimensión es 1, NumPy la "estira" virtualmente para igualarla a la otra (sin copiar memoria real).

Ejemplos típicos:

| Forma A    | Forma B    | ¿Compatible? | Resultado |
| ---------- | ---------- | ------------ | --------- |
| `(3, 4)`   | `(4,)`     | Sí           | `(3, 4)`  |
| `(3, 4)`   | `(3, 1)`   | Sí           | `(3, 4)`  |
| `(3, 4)`   | `(3,)`     | **No**       | error     |
| `(3, 1)`   | `(1, 4)`   | Sí           | `(3, 4)`  |

Uso típico en ciencia de datos: restar la media de cada columna a una matriz de datos (estandarización).

#### Mi ejemplo

Predice qué imprime antes de correr.

In [ ]:
import numpy as np

# Caso 1: matriz + vector fila (se aplica a cada fila)
M = np.array([
    [10, 20, 30],
    [40, 50, 60],
    [70, 80, 90],
])
fila = np.array([1, 2, 3])
print(M + fila)        # suma fila a CADA fila de M

print("---")

# Caso 2: matriz + vector columna (con shape (3,1))
columna = np.array([[100], [200], [300]])    # shape (3, 1)
print(M + columna)     # suma a cada COLUMNA

print("---")

# Caso típico: restar la media de cada columna a una matriz
datos = np.array([
    [1.0, 100.0],
    [2.0, 200.0],
    [3.0, 300.0],
    [4.0, 400.0],
])
medias = datos.mean(axis=0)    # shape (2,)
print("Medias:", medias)
print("Centrado:")
print(datos - medias)          # broadcasting: se resta a cada fila

[[11 22 33]
 [41 52 63]
 [71 82 93]]
---
[[110 120 130]
 [240 250 260]
 [370 380 390]]
---
Medias: [  2.5 250. ]
Centrado:
[[  -1.5 -150. ]
 [  -0.5  -50. ]
 [   0.5   50. ]
 [   1.5  150. ]]


#### Tu ejercicio espejo

En la celda de abajo:

1. Crea una matriz `M = np.arange(1, 13).reshape(3, 4)` y un vector `v = np.array([10, 20, 30, 40])`. Calcula `M + v`. Predice antes el resultado (¿se suma a las filas o a las columnas?).
2. Crea un vector columna `c = np.array([[100], [200], [300]])` (forma `(3, 1)`) y calcula `M + c`. ¿En qué dirección se aplica?
3. Crea una **tabla de multiplicar** del 1 al 10 usando broadcasting puro. Pista: `i = np.arange(1, 11).reshape(10, 1)` y `j = np.arange(1, 11).reshape(1, 10)`. Luego `i * j`. Imprime el resultado.
4. Dada `datos = np.array([[5, 100], [6, 110], [7, 120], [8, 130]])`, calcula la media de cada columna y **resta esa media a cada fila** (centrar los datos en cero). Imprime el resultado y verifica que las medias de las nuevas columnas son ~0.
5. **Bonus:** estandariza los datos del punto 4 (resta la media **y** divide por la desviación típica de cada columna). Esperado: cada columna debe tener media 0 y std 1.

**Pista para el punto 5:** `(datos - medias) / desviaciones`. Las dos formas tienen que ser `(2,)`.

In [ ]:
import numpy as np

# Tu solución aquí.


#### Tu reflexión

Completa la frase con tus palabras (doble click para editar):

> **"Broadcasting permite operar entre arrays de ___ distinta sin escribir bucles. Las formas son compatibles si, al alinearlas por la derecha, en cada dimensión son ___ o ___."**

---

## Nivel 6 — Herramientas para problemas reales

Cierre del notebook: tres herramientas que vas a usar en cuanto te enfrentes a un problema de datos serio. `np.where` para filtrar/sustituir, `np.random` para generar datos o muestrear, y un ejercicio integrador final.

### Bloque 15 — `np.where` y `np.random` (sustitución condicional y aleatoriedad)

**Concepto:** dos herramientas que aparecen en casi cualquier proyecto real.

**`np.where(condicion, valor_si_true, valor_si_false)`** es como un `if/else` vectorizado: recorre el array y, en cada posición, devuelve uno u otro valor según se cumpla la condición. Muy útil para limpiar datos.

**`np.random`** te genera datos aleatorios. Las funciones más usadas:

| Función                       | Qué genera                                                          |
| ----------------------------- | ------------------------------------------------------------------- |
| `np.random.rand(d1, d2, ...)` | Array con valores uniformes entre 0 y 1.                            |
| `np.random.randn(d1, d2, ...)`| Array con valores de una **normal estándar** (media 0, std 1).      |
| `np.random.randint(a, b, size)` | Enteros entre `a` (inclusivo) y `b` (exclusivo).                  |
| `np.random.seed(n)`           | Fija la semilla → resultados reproducibles entre ejecuciones.       |

**Truco fundamental:** llama a `np.random.seed(...)` antes de generar datos si quieres que tu compañero (o tu yo del futuro) obtenga **exactamente los mismos números** al reejecutar.

#### Mi ejemplo

Predice qué imprime antes de correr.

In [ ]:
import numpy as np

# np.where: limpiar negativos (sustituirlos por 0)
datos = np.array([3, -1, 5, -7, 2, 0, -4])
limpios = np.where(datos < 0, 0, datos)
print(datos)
print(limpios)

print("---")

# np.where: clasificar valores en categorías
notas = np.array([4, 7, 9, 3, 5, 8])
estado = np.where(notas >= 5, "Aprobado", "Suspenso")
print(estado)

print("---")

# np.random con seed: reproducible
np.random.seed(42)
print(np.random.rand(3))           # 3 floats uniformes [0, 1)
print(np.random.randint(0, 10, 5)) # 5 enteros entre 0 y 9
print(np.random.randn(2, 3))       # 2x3 de una normal estándar

[ 3 -1  5 -7  2  0 -4]
[3 0 5 0 2 0 0]
---
['Suspenso' 'Aprobado' 'Aprobado' 'Suspenso' 'Aprobado' 'Aprobado']
---
[0.37454012 0.95071431 0.73199394]
[4 6 9 2 6]
[[ 1.57921282  0.76743473 -0.46947439]
 [ 0.54256004 -0.46341769 -0.46572975]]


#### Tu ejercicio espejo

En la celda de abajo:

1. Crea `temperaturas = np.array([22, -5, 18, np.nan, 25, -3, 30])`. Usa `np.where` con `np.isnan` para sustituir los `NaN` por 0. Imprime el resultado.
2. Dado el mismo array (sin NaN ya), usa `np.where` para crear un array de strings que diga `"Frío"` si la temperatura es menor a 15 y `"Templado"` en caso contrario. Imprime.
3. Fija la semilla con `np.random.seed(0)` y genera una matriz `aleatoria` de 4x4 con `np.random.randint(0, 100, (4, 4))`. Imprime.
4. De `aleatoria`, calcula la **media de cada columna** (usa `axis`).
5. Usa `np.where` para crear una matriz `clasificada` con `"alto"` si el valor de `aleatoria` es mayor que su media de columna, y `"bajo"` si no. (Esta combinación broadcasting + where es exactamente la que se usa para detectar outliers en datos reales.) Pista: `np.where(aleatoria > aleatoria.mean(axis=0), "alto", "bajo")`.

**Pista importante para el punto 5:** la media de columna tiene forma `(4,)` y la matriz tiene forma `(4, 4)`. Broadcasting hace su magia.

In [ ]:
import numpy as np

# Tu solución aquí.


#### Tu reflexión

Completa la frase con tus palabras (doble click para editar):

> **"`np.where(cond, a, b)` actúa como un ___ vectorizado: devuelve `a` donde la condición es ___ y `b` donde es ___. `np.random.seed(n)` lo uso para que los números aleatorios sean ___ entre ejecuciones."**

### Bloque 16 — Ejercicio integrador: estandarizar una matriz y detectar outliers

**Concepto:** este bloque junta casi todo lo que viste: crear datos, atributos, reducciones por eje, broadcasting, máscaras booleanas y `np.where`. Es el tipo de ejercicio que verás constantemente en el máster.

**El problema a resolver:**

Tienes una matriz de medidas (por ejemplo, varias mediciones de varios sensores). Quieres:

1. Generar la matriz con datos aleatorios reproducibles.
2. Calcular media y desviación típica de cada columna.
3. **Estandarizar** (centrar y escalar): a cada valor le restas la media de su columna y lo divides por la desviación. Cada columna queda con media 0 y std 1.
4. Detectar **outliers**: valores que, tras estandarizar, están a más de 2 desviaciones de la media (es decir, `|valor| > 2`).
5. Sustituir los outliers por la media (que, después de estandarizar, es 0).

Esto se llama "z-score normalization" y es la primera técnica que aprendes en cualquier curso de preprocesado de datos.

#### Mi ejemplo

Lee el código con calma. Predice cómo será la salida (no los valores exactos, sino la **estructura**: una matriz, otra matriz, un escalar...).

In [ ]:
import numpy as np

# 1. Datos aleatorios reproducibles: 5 mediciones x 3 sensores
np.random.seed(7)
datos = np.random.randn(5, 3) * 10 + 50   # normal centrada en 50, std 10
print("Datos originales:")
print(datos)
print("Forma:", datos.shape)

print("---")

# 2. Media y desviación de CADA SENSOR (cada columna)
medias = datos.mean(axis=0)
stds   = datos.std(axis=0)
print("Media por sensor:", medias)
print("Std por sensor:  ", stds)

print("---")

# 3. Estandarización vectorizada (gracias a broadcasting)
estandar = (datos - medias) / stds
print("Datos estandarizados:")
print(estandar)
print("Medias tras estandarizar:", estandar.mean(axis=0).round(2))
print("Stds tras estandarizar:  ", estandar.std(axis=0).round(2))

print("---")

# 4. Detectar outliers: |z-score| > 2
outliers = np.abs(estandar) > 2
print("Máscara de outliers:")
print(outliers)
print("Número total de outliers:", outliers.sum())

print("---")

# 5. Sustituir outliers por 0 (la media en el espacio estandarizado)
limpio = np.where(outliers, 0, estandar)
print("Datos limpios:")
print(limpio)

Datos originales:
[[66.90525704 45.34062629 50.32820164]
 [54.07516283 42.11076971 50.02065573]
 [49.99109614 32.45275694 60.17658006]
 [56.00498516 43.74571026 48.28451739]
 [55.05299374 47.38643585 47.57250921]]
Forma: (5, 3)
---
Media por sensor: [56.40589898 42.20725981 51.2764928 ]
Std por sensor:   [5.63577201 5.18012194 4.56857412]
---
Datos estandarizados:
[[ 1.86298488  0.60488277 -0.2075683 ]
 [-0.41356111 -0.01862699 -0.274886  ]
 [-1.13822966 -1.88306434  1.94811051]
 [-0.07113734  0.29699117 -0.65490355]
 [-0.24005677  0.9998174  -0.81075265]]
Medias tras estandarizar: [ 0. -0.  0.]
Stds tras estandarizar:   [1. 1. 1.]
---
Máscara de outliers:
[[False False False]
 [False False False]
 [False False False]
 [False False False]
 [False False False]]
Número total de outliers: 0
---
Datos limpios:
[[ 1.86298488  0.60488277 -0.2075683 ]
 [-0.41356111 -0.01862699 -0.274886  ]
 [-1.13822966 -1.88306434  1.94811051]
 [-0.07113734  0.29699117 -0.65490355]
 [-0.24005677  0.9998174  

#### Tu ejercicio espejo

Versión libre: **vas a hacer lo mismo, con tus datos y tomando las decisiones tú**.

En la celda de abajo:

1. Fija la semilla con `np.random.seed(123)` (un número distinto al mío, para tener otros datos).
2. Genera una matriz `notas_clase` de **20 alumnos x 4 asignaturas** con `np.random.randint(0, 11, (20, 4))` (notas enteras del 0 al 10).
3. Calcula e imprime:
   - La nota media de cada **asignatura** (4 valores).
   - La nota media de cada **alumno** (20 valores).
   - El número de **suspensos por asignatura** (notas < 5, contadas por columna).
   - El número de **alumnos que aprobaron todo** (todas sus 4 notas >= 5). Pista: `(notas_clase >= 5).all(axis=1).sum()`.
4. Estandariza la matriz **por columna** (cada asignatura tendrá media 0 y std 1).
5. Usa `np.where` para crear una matriz `categorias` donde cada valor estandarizado se transforma en:
   - `"alto"` si el z-score > 1
   - `"bajo"` si el z-score < -1
   - `"medio"` en cualquier otro caso

**Pista para el punto 5:** `np.where` solo tiene dos ramas, así que tendrás que **anidarlo**: `np.where(cond1, valor1, np.where(cond2, valor2, valor3))`.

**Si haces este bloque entero**, ya manejas NumPy a un nivel suficiente para entender cualquier código de scikit-learn o Pandas que veas en el resto del máster.

In [ ]:
import numpy as np

# Tu solución aquí.


#### Tu reflexión

Completa la frase con tus palabras (doble click para editar):

> **"Para estandarizar (z-score) una matriz por columnas hago `(datos - ___) / ___`, donde ambas se calculan con `axis=___`. Esto funciona gracias a ___, que estira la forma del vector hasta encajar con la matriz."**

---

## Cierre

Si llegaste hasta aquí y resolviste tú mismo los 16 espejos, ya tienes la base sólida de NumPy. No es una base completa — quedan álgebra lineal avanzada, manipulación de archivos `.npy`, broadcasting de más de 2 dimensiones, etc. — pero **es la base sobre la que se construye el 90% del código de ciencia de datos que verás**.

Si en algún bloque te quedaste atascado, vuelve. Releer el concepto con los datos del espejo bajo el brazo es de las cosas más útiles que puedes hacer.

Buena suerte con el resto del MIAR.